### PR TITLE
S3 settings functional tests

### PR DESCRIPTION
**Pull request by osulzhenko**

### CODE REVIEW

**Comment by SerhiiNahornyi** [comment]
@muuki88 We added a few tests for your PR
https://github.com/prebid/prebid-server-java/pull/2837

Could you pls review them with us?

**Comment by muuki88** [comment]
Sure thing 😃 I'll try to review ASAP

File: src/test/groovy/org/prebid/server/functional/tests/storage/AccountS3Spec.groovy [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+        when: "PBS processes auction request"
+        s3StorageAccountPbsService.sendAuctionRequest(bidRequest)
+
+        then: "PBS should reject the entire auction"
+        def exception = thrown(PrebidServerException)
+        assert exception.statusCode == UNAUTHORIZED.code()
+        assert exception.responseBody == "Account $accountId is inactive"
+    }
+
+    def "PBS should throw exception when account have different id inside json in S3 storage"() {
```
Comment (muuki88) [inline src/test/groovy/org/prebid/server/functional/tests/storage/AccountS3Spec.groovy:62]: This test makes sense and the implementation is missing this check

File: src/test/groovy/org/prebid/server/functional/tests/storage/AmpS3Spec.groovy [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+        assert bidderRequest.site?.page == ampStoredRequest.site.page
+        assert bidderRequest.site?.publisher?.id == ampStoredRequest.site.publisher.id
+        assert !bidderRequest.imp[0]?.tagId
+        assert bidderRequest.imp[0]?.banner?.format[0]?.h == ampStoredRequest.imp[0].banner.format[0].h
+        assert bidderRequest.imp[0]?.banner?.format[0]?.w == ampStoredRequest.imp[0].banner.format[0].w
+        assert bidderRequest.regs?.gdpr == ampStoredRequest.regs.ext.gdpr
+    }
+
+    def "PBS should throw exception when trying to take parameters from the stored request on S3 service with invalid id in file"() {
```
Comment (muuki88) [inline src/test/groovy/org/prebid/server/functional/tests/storage/AmpS3Spec.groovy:45]: This test makes sense and currently fails in the pipeline.

File: src/test/groovy/org/prebid/server/functional/tests/storage/AuctionS3Spec.groovy [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+        when: "Requesting PBS auction"
+        s3StoragePbsService.sendAuctionRequest(bidRequest)
+
+        then: "Response should contain imp[0].secure same value as in request"
+        def bidderRequest = bidder.getBidderRequest(bidRequest.id)
+        assert bidderRequest.imp[0].secure == secureStoredRequest
+    }
+
+    def "PBS should throw exception when trying populate imp[0].secure from imp stored request on S3 service with invalid impId in file"() {
```
Comment (muuki88) [inline src/test/groovy/org/prebid/server/functional/tests/storage/AuctionS3Spec.groovy:42]:
```suggestion
    def "PBS should throw exception when trying to populate imp[0].secure from imp stored request on S3 service with invalid impId in file"() {
```

This test I don't understand. What is in `imp[0].secure` that this fails the test? Is this another layer of checksum/validation to ensure the integrity of the configuration?

File: src/test/groovy/org/prebid/server/functional/tests/storage/StoredResponseS3Spec.groovy [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        assert response.seatbid[0]?.bid?.size() == storedAuctionResponse.bid.size()
+        assert response.seatbid[0]?.bid[0]?.impid == storedAuctionResponse.bid[0].impid
+        assert response.seatbid[0]?.bid[0]?.price == storedAuctionResponse.bid[0].price
+        assert response.seatbid[0]?.bid[0]?.id == storedAuctionResponse.bid[0].id
+
+        and: "PBS not send request to bidder"
+        assert !bidder.getRequestCount(bidRequest.id)
+    }
+
+    def "PBS should throw request format exception when stored auction response have different id inside S3 file"() {
```
Comment (muuki88) [inline src/test/groovy/org/prebid/server/functional/tests/storage/StoredResponseS3Spec.groovy:42]: This is probably sane to do that filename and ID match.

**Review by muuki88 — APPROVED** [review]
First of all: kudos to the structure of your functional tests. Super comfortable to read and understand. Even though most of them look like they could be unit tests :smile: 

3 of the 4 failing tests make totally sense. The one with `.secure` property I don't get, because I'm not familiar with this feature and what the intention is.

My question is: Do you have the capacity to add the missing checks to make the tests green?

File: src/test/groovy/org/prebid/server/functional/tests/storage/AuctionS3Spec.groovy (hunk 3 above)
Comment (muuki88) [reply src/test/groovy/org/prebid/server/functional/tests/storage/AuctionS3Spec.groovy:42]: Okay, got it: https://docs.prebid.org/prebid-server/endpoints/openrtb2/pbs-endpoint-auction.html#determining-bid-security-httphttps 

it's a simple flag, forcing https or not

File: src/test/groovy/org/prebid/server/functional/tests/storage/StorageBaseSpec.groovy [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+import org.prebid.server.functional.testcontainers.PbsServiceFactory
+import org.prebid.server.functional.tests.BaseSpec
+import org.prebid.server.functional.util.PBSUtils
+
+class StorageBaseSpec extends BaseSpec {
+
+    protected static final String INVALID_FILE_BODY = 'INVALID'
+    protected static final def DEFAULT_BUCKET = PBSUtils.randomString.toLowerCase()
+
+    protected static final S3Service S3_SERVICE = new S3Service(DEFAULT_BUCKET)
```
Comment (marki1an) [inline src/test/groovy/org/prebid/server/functional/tests/storage/StorageBaseSpec.groovy:14]: Looks weard when on constant we call method like here `S3_SERVICE.uploadStoredResponse(...)`
As for me would be great move to lower case `s3Service`

File: src/test/groovy/org/prebid/server/functional/tests/storage/StorageBaseSpec.groovy [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+import org.prebid.server.functional.service.PrebidServerService
+import org.prebid.server.functional.service.S3Service
+import org.prebid.server.functional.testcontainers.PbsServiceFactory
+import org.prebid.server.functional.tests.BaseSpec
+import org.prebid.server.functional.util.PBSUtils
+
+class StorageBaseSpec extends BaseSpec {
+
+    protected static final String INVALID_FILE_BODY = 'INVALID'
+    protected static final def DEFAULT_BUCKET = PBSUtils.randomString.toLowerCase()
```
Comment (marki1an) [inline src/test/groovy/org/prebid/server/functional/tests/storage/StorageBaseSpec.groovy:12]: Pretty minor, for consistency move `def` to `String`

File: src/test/groovy/org/prebid/server/functional/tests/storage/StorageBaseSpec.groovy [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    protected static Map<String, String> s3StorageConfig = [
+            'settings.s3.accessKeyId'         : "${S3_SERVICE.accessKeyId}".toString(),
+            'settings.s3.secretAccessKey'     : "${S3_SERVICE.secretKeyId}".toString(),
+            'settings.s3.endpoint'            : "${S3_SERVICE.endpoint}".toString(),
+            'settings.s3.bucket'              : "${DEFAULT_BUCKET}".toString(),
+            'settings.s3.accounts-dir'        : "${S3Service.DEFAULT_ACCOUNT_DIR}".toString(),
+            'settings.s3.stored-imps-dir'     : "${S3Service.DEFAULT_IMPS_DIR}".toString(),
+            'settings.s3.stored-requests-dir' : "${S3Service.DEFAULT_REQUEST_DIR}".toString(),
+            'settings.s3.stored-responses-dir': "${S3Service.DEFAULT_RESPONSE_DIR}".toString(),
+    ]
```
Comment (marki1an) [inline src/test/groovy/org/prebid/server/functional/tests/storage/StorageBaseSpec.groovy:25]: Please remove redundant `.toString()`
```suggestion
     protected static Map<String, String> s3StorageConfig = [
            'settings.s3.accessKeyId'         : s3Service.accessKeyId,
            'settings.s3.secretAccessKey'     : s3Service.secretKeyId,
            'settings.s3.endpoint'            : s3Service.endpoint,
            'settings.s3.bucket'              : DEFAULT_BUCKET,
            'settings.s3.accounts-dir'        : S3Service.DEFAULT_ACCOUNT_DIR,
            'settings.s3.stored-imps-dir'     : S3Service.DEFAULT_IMPS_DIR,
            'settings.s3.stored-requests-dir' : S3Service.DEFAULT_IMPS_DIR,
            'settings.s3.stored-responses-dir': S3Service.DEFAULT_IMPS_DIR,
    ]
```

File: src/test/groovy/org/prebid/server/functional/tests/storage/AccountS3Spec.groovy (hunk 1 above)
Comment (marki1an) [inline src/test/groovy/org/prebid/server/functional/tests/storage/AccountS3Spec.groovy:62]: `PBS should throw exception when account id is't match wiht bid request account id`
Please check other similar case

**Review by marki1an — CHANGES_REQUESTED** [review]

**Comment by osulzhenko** [comment]
> First of all: kudos to the structure of your functional tests. Super comfortable to read and understand. Even though most of them look like they could be unit tests 😄
> 
> 3 of the 4 failing tests make totally sense. The one with `.secure` property I don't get, because I'm not familiar with this feature and what the intention is.
> 
> My question is: Do you have the capacity to add the missing checks to make the tests green?

@muuki88  It’s possible to add the missing checks here to make the tests pass. However, for more consistency, I suggest adding these checks in your Pull Request (PR). Once done, I can merge them with the current PR.

File: src/test/groovy/org/prebid/server/functional/tests/storage/AccountS3Spec.groovy (hunk 1 above)
Comment (osulzhenko) [reply src/test/groovy/org/prebid/server/functional/tests/storage/AccountS3Spec.groovy]: done

File: src/test/groovy/org/prebid/server/functional/tests/storage/StorageBaseSpec.groovy (hunk 7 above)
Comment (osulzhenko) [reply src/test/groovy/org/prebid/server/functional/tests/storage/StorageBaseSpec.groovy]: done

File: src/test/groovy/org/prebid/server/functional/tests/storage/StorageBaseSpec.groovy (hunk 6 above)
Comment (osulzhenko) [reply src/test/groovy/org/prebid/server/functional/tests/storage/StorageBaseSpec.groovy]: done

File: src/test/groovy/org/prebid/server/functional/tests/storage/StorageBaseSpec.groovy (hunk 5 above)
Comment (osulzhenko) [reply src/test/groovy/org/prebid/server/functional/tests/storage/StorageBaseSpec.groovy]: done

File: src/test/groovy/org/prebid/server/functional/service/S3Service.groovy [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+                .withNetwork(Dependencies.network)
+                .withServices(LocalStackContainer.Service.S3)
+        localStackContainer.start()
+        s3PbsService = S3Client.builder()
+                .endpointOverride(localStackContainer.getEndpointOverride(LocalStackContainer.Service.S3))
+                .credentialsProvider(StaticCredentialsProvider.create(AwsBasicCredentials.create(
+                        localStackContainer.getAccessKey(), localStackContainer.getSecretKey())))
+                .region(Region.of(localStackContainer.getRegion()))
+                .build()
+        createBucket(bucketName)
```
[the remark this answers is not in the document]
Comment (osulzhenko) [reply src/test/groovy/org/prebid/server/functional/service/S3Service.groovy]: done

File: src/test/groovy/org/prebid/server/functional/service/S3Service.groovy [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+    static final def DEFAULT_ACCOUNT_DIR = 'account'
+    static final def DEFAULT_IMPS_DIR = 'stored-impressions'
+    static final def DEFAULT_REQUEST_DIR = 'stored-requests'
+    static final def DEFAULT_RESPONSE_DIR = 'stored-responses'
+
+    S3Service(String bucketName) {
+        this.localStackContainer = new LocalStackContainer(DockerImageName.parse("localstack/localstack:s3-latest"))
+                .withNetwork(Dependencies.network)
+                .withServices(LocalStackContainer.Service.S3)
+        localStackContainer.start()
```
[the remark this answers is not in the document]
Comment (osulzhenko) [reply src/test/groovy/org/prebid/server/functional/service/S3Service.groovy]: done

**Review by marki1an — DISMISSED** [review]

File: src/test/groovy/org/prebid/server/functional/tests/storage/AccountS3Spec.groovy (hunk 1 above)
Comment (muuki88) [reply src/test/groovy/org/prebid/server/functional/tests/storage/AccountS3Spec.groovy]: Done

**Comment by muuki88** [comment]
@SerhiiNahornyi , @marki1an I need to push back on some of the failing tests, as they are currently very hard to implement.

The [StoredDataResult](https://github.com/prebid/prebid-server-java/blob/master/src/main/java/org/prebid/server/settings/model/StoredDataResult.java) model contains the results untyped. So this validation is currently not available in _any_ `ApplicationSettings` implementation (FilesystemSettings, JdbcSettings).

I would argue that these kind of validations should be done in a generic way a level higher. Could you remove the failing tests?

**Comment by marki1an** [comment]
> @SerhiiNahornyi , @marki1an I need to push back on some of the failing tests, as they are currently very hard to implement.
> 
> The [StoredDataResult](https://github.com/prebid/prebid-server-java/blob/master/src/main/java/org/prebid/server/settings/model/StoredDataResult.java) model contains the results untyped. So this validation is currently not available in _any_ `ApplicationSettings` implementation (FilesystemSettings, JdbcSettings).
> 
> I would argue that these kind of validations should be done in a generic way a level higher. Could you remove the failing tests?

@muuki88 Hi there.
Sure, we'll mark those tests with `@PedningFeature` annotation, and those tests will skip while annotation present.
@osulzhenko Can you mark those tests?

**Comment by osulzhenko** [comment]
> > @SerhiiNahornyi , @marki1an I need to push back on some of the failing tests, as they are currently very hard to implement.
> > The [StoredDataResult](https://github.com/prebid/prebid-server-java/blob/master/src/main/java/org/prebid/server/settings/model/StoredDataResult.java) model contains the results untyped. So this validation is currently not available in _any_ `ApplicationSettings` implementation (FilesystemSettings, JdbcSettings).
> > I would argue that these kind of validations should be done in a generic way a level higher. Could you remove the failing tests?
> 
> @muuki88 Hi there. Sure, we'll mark those tests with `@PedningFeature` annotation, and those tests will skip while annotation present. @osulzhenko Can you mark those tests?

@osulzhenko sure, marked

**Comment by muuki88** [comment]
@osulzhenko anything that blocks this?

**Comment by osulzhenko** [comment]
> @osulzhenko anything that blocks this?

@muuki88 Not from my side. You need to get approval on your original PR https://github.com/prebid/prebid-server-java/pull/2733

**Comment by Net-burst** [comment]
The workflow still fails on S3 unit tests:
```
Error:  Errors: 
Error:    S3ApplicationSettingsTest.getStoredDataReturnsErrorsForNotFoundImpressions:359 » Sdk
Error:    S3ApplicationSettingsTest.getStoredDataReturnsErrorsForNotFoundRequests:329 » Sdk
Error:    S3ApplicationSettingsTest.getStoredDataShouldReturnFetchedStoredImpression » ClassCast
Error:    S3ApplicationSettingsTest.getStoredDataShouldReturnFetchedStoredImpressionAndStoredRequest » ClassCast
Error:    S3ApplicationSettingsTest.getStoredDataShouldReturnFetchedStoredImpressionWithAdUnitPathStoredId » ClassCast
Error:    S3ApplicationSettingsTest.getStoredDataShouldReturnFetchedStoredRequest » ClassCast
[INFO] 
Error:  Tests run: 7271, Failures: 0, Errors: 6, Skipped: 0
```

**Review by marki1an — APPROVED** [review]

**Pull request merged**